PLANET-5017: Re-implement Spreadsheet block frontend in React

pablocubico opened the pull request:
Porting the Spreadsheet block's frontend to React.

Ref: https://jira.greenpeace.org/browse/PLANET-5017

pablocubico commented:
A few notes about this:

0. I need to squash some commits!

1. I tried to avoid the `render_callback` part on the PHP side by just rendering the node in which we inject the React component in the `save()` method of the editor, but using the `save()` method in the Block definition triggers a validation error in the Editor, as our current blocks return `null`, any value different than `null` will trigger a validation error and ask you to "Resolve" the block (yuck!). We could just return an HTML node there to render the React content in the frontend, but changing the `save()` function will make the current spreadsheets unusable. I just made it a "template-less" render, it just renders a node. Ideally I would love to skip the PHP side in these cases, but let's stick to `null` and `render_callback` for now. ( This is a long standing controversy in Gutenberg, there are a few discussions about it out there:
https://github.com/WordPress/gutenberg/issues/10444 )

2. I added an endpoint to transform the CSV to JSON and keep using WP's cache store the Spreadsheet as JSON data. I was thinking it would be fast since the cache is persistent, so the data should be obtained from cache even from different visitors, but then @Inwerpsel came in and said: "Hey, why don't you just get the spreadsheet from the client side?", which is pretty much the same from the component's point of view, is just a bunch of JSON rows. So, as Pieter said, that could avoid WP's overhead, but I guess it could add some overhead to the Javascript side since we need to turn that CSV into something iterable, which is super simple but if the Spreadsheet is long, it may have a performance impact on low-end cellphones. Which is also a concern for React, the React embed is not that big, but it adds quite a bunch of logic for the browser to process, despite the KB size. There is also some cost in the DNS Lookup and connecting to Google if we load it directly from the client side. 

So, I'm not sure about (2), maybe we can test these assumptions? Is the WP overhead that bad? Is parsing the CSV lines to JSON super fast? What do you think? 

Ping @Inwerpsel @sagarsdeshmukh @comzeradd

comzeradd commented:
Regarding point 2: I'm leaning mostly towards 1st option. Since we already need to render the block in the backend, it seems less complicated and faster for the end-user to use WP cache store. Unless I'm missing something, it also feels more work to do the 2nd option. Which is not necessarily bad if that's the proper way of implementation. But maybe this block is a good opportunity to see how this works and do something similar with WYSIWYG blocks. Rendering on the client-side could be a very good option if we didn't render the block in the backend at all (eg. the native Embed block for Facebook).

Inwerpsel commented:
Regarding 1. Seems like the most sensible thing to do indeed. Maybe we can set this up in a generic way for all blocks that are rendered frontend? We could use
```
<div data-block="{{ block.id }}" data-attributes="{{ block.attributes.toJson }}" class="{{ block.placeholder_class }}"></div>
```
for all blocks.

Actually, maybe then we can also circumvent the `save()` issue :thinking: Because we would have only one "block shape change" to deal with, from null to this format, which we can write a resolver for, probably. We would only need to write one as the element and attributes don't need to change to support other blocks. Even if the html would be changed externally, we can write a resolver for that as the content can be entirely reconstructed from the block attributes in the html comment.

Maybe worth a shot? Then we wouldn't need the render on the backend, which is basically just turning one representation of the block's data into another one (block comment -> data attributes).

As for 2. I think the CSV parsing should be really fast and only an issue for huge spreadsheets, like 100k or more rows, far beyond what would make sense to include on a page. But we can try if we want to be sure. Possibly we can get this data as JSON too from the Google sheets API.

The reason I remarked it was because I assume that Google sheets has a pretty extensive CDN to deliver this data efficiently anywhere in the world, so very likely the round trip to our server will be significantly longer, even without WP overhead. Note that currently even though a placeholder for the blocks is rendered on the backend, the data of the sheet doesn't come with that. The frontend needs to do a separate call to a custom endpoint before it can start rendering. So for the end user the difference in speed is between how fast we can serve a static resource vs how fast google can do that.

Though serving it from our backend also has some advantages, as it ensures the sheet will work regardless of possible issues that might prevent the browser from getting the data from Google. There might be some other considerations when including data from external domains.

I'm ok with the current approach of caching it on our server as there's probably no noticeable difference to the end user in most cases. Probably the additional load on our servers is also negligible.

pablocubico commented:
Hey @Inwerpsel , sorry for the delay.

Regarding 1:
I tried doing that in the `save()` method but my version was using `data-url`, a "per attribute" property, which does not scale well if we change the attributes..., it actually makes sense if to only pass "data-attributes" to it and extract the ones we want in the frontend renderer. I'll give it a try and use the "deprecations" feature of Gutenberg to bypass the validation issue.

Regarding 2:
Alright, let's keep the current version which uses the cache and could help in avoiding some DNS latency or Google issues... in case Google has any.  We can always change it to a client fetch since it's pretty much the same from the frontend side.

pablocubico commented:
@comzeradd @Inwerpsel 

Branch updated to latest master. I removed the `render_callback` from the PHP side and created a generic method to render blocks in the frontend. The `save()` provides a `className` which includes the block's name but also any additional classes from the [block styles](https://www.billerickson.net/block-styles-in-gutenberg/) feature, so it's quite handy.

I wrote a deprecation handler to avoid any warnings from the `save()` method change. Looks pretty good. I also ported the CSS variables from the spreadsheet colors feature from Twig to React.

Inwerpsel commented on assets/src/frontendIndex.js:
```diff
@@ -0,0 +1,20 @@
+
+const getFrontendBlockComponent = blockNode => COMPONENTS[
+	Array.from(blockNode.classList)
+		.find( className => className.includes( PLANET4_BLOCK_PREFIX ))
```
Maybe we can create a dedicated `data-` attribute for this instead of a class? Something like `data-render="spreadsheet"` Then we could get the block name from the attribute's value, instead of having to search the classnames and remove the prefix. [The querySelector below](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/273/files?file-filters%5B%5D=.js&file-filters%5B%5D=.php&file-filters%5B%5D=.scss&hide-deleted-files=true#diff-e318476ddd94efc2f71618da8e052ddbR14) would then also be a bit simpler, just matching any elements that have the data attribute, regardless of its value.
```
document.querySelectorAll(`[data-render]`).forEach(...)
```

pablocubico replied:
That was the initial approach, but then I couldn't find a way to get the block name in the `save()` method, it receives the `attributes` and a set of `className`s which includes the block's class and any "block style" classes. But I guess I can just make the block name a constant in the block registration constructor and use it in `save()`, and make it a prop of the `FrontendNode`. I think I'll also include the `classNames` (which I'm not doing now) so we can use block styles in the future (or any other CSS stuff).

Inwerpsel commented on classes/blocks/class-spreadsheet.php:
```diff
@@ -49,9 +48,6 @@ public function __construct() {
 	 * @return array The data to be passed in the View.
 	 */
 	public function prepare_data( $fields ): array {
```
If `prepare_data` is used only in the backend render, then you can remove this function, or make it just return an empty array.

pablocubico commented on assets/src/blocks/Spreadsheet/SpreadsheetFrontend.js:
```diff
@@ -0,0 +1,186 @@
+      invalidSheetId: false,
+      searchText: '',
+      sortDirection: 'asc',
+      sortColumnIndex: 0,
```
I'll make it `null` then and leave it unsorted at start. 👍

pablocubico commented on assets/src/blocks/Spreadsheet/SpreadsheetFrontend.js:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  onHeaderClick( index ) {
+    if ( index == this.state.sortColumnIndex ) {
```
Hm I think so, I'll remove it and test.

pablocubico replied:
So it turns out this is actually needed to set the sortDirection to `asc` in case you clicked in a different header, otherwise it could be `desc` and trigger the animation on click. Anyway, Magali and I will review the header behavior as the next iteration, so this will probably need some changes.

pablocubico commented on assets/src/blocks/Spreadsheet/SpreadsheetFrontend.js:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  highlightMatches( cellValue, searchText ) {
+    let reg = new RegExp('(' + searchText + ')', 'gi')
```
Actually that was my first suggestion when we estimated PLANET-4914 on a backlog refinement call, but you suggested doing it ourselves to avoid including a library as it would be pretty simple... which actually made sense, I just did a quick test adding the library and the `frontendIndex` build went from 7.2K to 24K... I think we can keep this version unless we need extra features as the ones provided by that library or we start using highlighting in other places...

Inwerpsel replied:
Right, I did suggest that :) Yeah you're right, that increase in frontend footprint doesn't seem worth it if it's our only use case. Still would be nicer if we extracted the highlighting logic from `SpreadsheetFrontend` so we can keep that one a bit more high level.

pablocubico commented on assets/src/blocks/Spreadsheet/Spreadsheet.js:
```diff
@@ -89,20 +89,18 @@ export class Spreadsheet extends Component {
     );
   }
 
+  renderView() {
```
Hmmm not sure I follow, this is not rendering the placeholder as we do in the frontend... maybe you saw it's not rendering, the reason the table is not being rendered is because of a bug I introduced in the endpoint here: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/273/files#diff-765c584bd08bac4ddefd3a6bfc6e4706R126

I am validating the `sheet_id` as if it was an integer, I don't know why I thought so... just out of habit I guess...

pablocubico commented on assets/src/blocks/Spreadsheet/SpreadsheetFrontend.js:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  renderRows() {
+    const rows = this.filterMatchingRows( this.sortRows( this.state.sheet_data.rows, this.state.sortColumnIndex ) );
```
Good catch!! 💯

pablocubico commented on assets/src/blocks/Spreadsheet/Spreadsheet.js:
```diff
@@ -89,20 +89,18 @@ export class Spreadsheet extends Component {
           this.props.isSelected
             ? this.renderEdit()
-            : null
+            : this.renderView()
```
I would actually love to move the field to the sidebar as it's part of the WYSIWYG refactors doc, most fields in our blocks should be in the sidebar if we want to provide a WYSIWYG UX. I'll see what I can do in terms of UI there.

mleray commented on assets/src/blocks/Spreadsheet/Spreadsheet.js:
```diff
@@ -68,42 +68,60 @@ export class Spreadsheet extends Component {
+            <div className="sidebar-blocks-help">
+              <ul>
+                <li>
+                  { __(`From Your Google Spreadsheet Table choose File -> Publish on web.
```
The description has been updated for https://jira.greenpeace.org/browse/PLANET-5022, so I guess it should be updated here too?

pablocubico replied:
Good catch!

pablocubico commented:
Hey @Inwerpsel @comzeradd , branch rebased to master, this is currently deployed in the Mars environment.

Addressed all the comments in the review, applied most of the suggestions. I didn't decouple the rendering from the API fetch, I think we can do that later if the requirements change, for example, if we are planning to fetch spreadsheets from other sources than Google docs. 

I also considered registering a Redux Store in the Gutenberg side to handle the data, using the `withSelect/withDispatch/subscribe` combo there instead of `apiFetch`, that way we could use the Redux debugger to monitor the requests, even do rewind, play, and such, but it has no real benefit from the frontend side, as there are no Stores registered there, and seems overkill. For the moment, WP is still focused on the Editor side for their React stuff.

Inwerpsel commented on classes/blocks/class-spreadsheet.php:
```diff
@@ -42,49 +41,10 @@ public function __construct() {
-		}
-
-		return $id;
+		return $fields;
```
Probably makes more sense to return an empty array here, as this function normally is't supposed to return `$fields`, even though some of our blocks do that.

pablocubico commented on assets/src/blocks/CSSVariablesToText.js:
```diff
@@ -0,0 +1,4 @@
+export const CSSVariablesToText = CSSVariables =>
+	Object.entries( CSSVariables )
+		.map( cssVariable => `--${cssVariable[0]}: ${cssVariable[1]}` )
+		.join(';'); // React does not like an ending semicolon on cssText
```
React actually throws an error if there is a semicolon in the last one: `Warning: Style property values shouldn't contain a semicolon` and suggests removing it. I'll clarify my comment.

pablocubico commented on assets/src/blocks/CSSVariablesToText.js:
```diff
@@ -0,0 +1,4 @@
+export const CSSVariablesToText = CSSVariables =>
```
Hmmm I was try to imply this is meant to be inlined and used in the `cssText` property (https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/cssText) , I'll figure out a better name... 🤔

pablocubico commented on assets/src/blocks/Spreadsheet/SpreadsheetFrontend.js:
```diff
@@ -0,0 +1,228 @@
+              </thead>
+              <tbody>
+                {
+                  this.state.loading === true
```
This is a feature: there is an initial state where `loading`  is `null`, when there is no URL set yet, and it shows some bogus content in the Editor so you can actually see how the Spreadsheet would look like even if the URL is not set yet.

comzeradd commented on assets/src/blocks/Spreadsheet/SpreadsheetFrontend.js:
```diff
@@ -0,0 +1,217 @@
+
+    return (
+      <Fragment>
+        <div className="block-spreadsheet" style={{ cssText: CSSVariablesToInlineText( this.props.css_variables ) }}>
```
do you think it would be too complex to have the block inside a `<section>`, instead of a `<div>`. To be more semantically meaningful, but also to preserve consistency with existing blocks markup.

pablocubico replied:
Sure, forgot about that.

Inwerpsel commented on assets/src/blocks/CSSVariablesToInlineText.js:
```diff
@@ -0,0 +1,14 @@
+export const CSSVariablesToInlineText = CSSVariables => {
+	if ( ! CSSVariables ) return '';
+	return Object.entries( CSSVariables )
+		.map( cssVariable => `--${cssVariable[0]}: ${cssVariable[1]}` )
```
You can destructure the array here for better readability:
```
		.map( [ key, value ] => `--${key}: ${value}` )
```

Inwerpsel commented on assets/src/blocks/CSSVariablesToInlineText.js:
```diff
@@ -0,0 +1,14 @@
+ * @param {object} CSSVariables
+ * @returns {string}
+ */
+export const CSSVariablesToInlineText = CSSVariables => {
```
Sorry to continue on the name nit picking :) But the inline part is not really accurate, you could as write the result of this function to a css file with a `::root {}` selector. Maybe `toDeclarations = CssVariables => {}` makes sense here?

Inwerpsel commented on assets/src/blocks/Spreadsheet/SpreadsheetEditor.js:
```diff
@@ -0,0 +1,148 @@
+
+    return <Fragment>
+      {
+        ! this.props.attributes.url
```
`this.props.attributes` is used a few times here, maybe easier on the eye if we destructure it `const { attributes } = this.props`.

Inwerpsel commented on assets/src/blocks/HighlightMatches.js:
```diff
@@ -0,0 +1,17 @@
+import { Fragment } from '@wordpress/element';
+
+export const HighlightMatches = ( cellValue, searchText, className = 'highlighted-text' ) => {
+	let reg = new RegExp('(' + searchText.trim() + ')', 'gi')
```
Not sure if this is the right place to perform the `trim` on the searchtext. It's not impossible that we actually want to match a trailing whitespace. https://github.com/bvaughn/react-highlight-words also works that way, you can check at https://codesandbox.io/s/5v8yqoxv7k?file=/src/index.js

Inwerpsel replied:
And we can just trim `searchText` once when setting it on the state, shouldn't be needed afterwards.

pablocubico replied:
Trimming the value on the state actually prevents you from typing any spaces in the controlled input, is not impossible to match a trailing whitespace but I think it's an edge case, also, as this highlight implementation is just making the text bold, you can't actually see a space being highlighted, so I guess I'll just keep it as is for now, we can adopt that library afterwards if it makes sense... I think we will probably add it at some point...

Inwerpsel commented on assets/src/blocks/Spreadsheet/SpreadsheetFrontend.js:
```diff
@@ -0,0 +1,217 @@
+
+  filterMatchingRows( rows ) {
+    const filteredRows = rows.filter( row => {
+      return !!row.find( cell => cell.toLowerCase().includes( this.state.searchText.trim().toLowerCase() ) );
```
You can use `row.some` instead here, which already returns a boolean for you.

pablocubico replied:
Neat 💯

Inwerpsel commented on classes/blocks/class-spreadsheet.php:
```diff
@@ -42,49 +41,12 @@ public function __construct() {
-	 * @param array $fields This is the array of fields of this block.
-	 *
-	 * @return array The data to be passed in the View.
+	 * @param array $fields Unused, required by the abstract function.
```
I'm usually not a fan of long inheritance chains, but maybe it's more appropriate here if we create sub-types for each kind of block? So that would make
```
|Base_Block
|\
||Static_Block
|\
||Dynamic|Block
```

Inwerpsel replied:
Not proposing to do this in this PR btw :D probably best handled separately.

Inwerpsel replied:
Then we could just move the `prepare_data` function to `Dynamic_Block`.

pablocubico replied:
Indeed, this is calling for a few refactors.
